### PR TITLE
ENH: Add duck-typing namespace scraper.

### PR DESCRIPTION
**Pull request by danielballan**
This device does not get collected by `get_all_positioners`:

```python
class XYMotor(Device):
    x = Cpt(EpicsMotor, ...)
    y = Cpt(EpicsMotor, ...)

Should we add a `set` method and use duck-typing to identify positioners? See attached function.

### CODE REVIEW

File: ophyd/commands.py [hunk 1]
```diff
@@ -61,6 +61,26 @@ def instances_from_namespace(classes):
                 if isinstance(val, classes) and not var.startswith('_')]
 
 
+def ducks_from_namespace(attr):
```
Comment (tacaswell) [inline ophyd/commands.py]: This should take *args of attr names (with implicit `&` logic)

File: ophyd/commands.py [hunk 2]
```diff
@@ -392,17 +443,16 @@ def log_pos(positioners=None, extra_msg=None):
...
-    if positioners is not None:
-        msg += logbook_add_objects(positioners)
+    msg += logbook_add_objects(positioners)
 
-        for p in positioners:
-            try:
-                pdict['values'][p.name] = p.position
-            except DisconnectedError:
-                pdict['values'][p.name] = DISCONNECTED
+    for p in positioners:
```
Comment (tacaswell) [inline ophyd/commands.py]: This needs the `None` guard added back.

File: ophyd/commands.py [hunk 3]
```diff
@@ -51,19 +67,54 @@ def instances_from_namespace(classes):
...
+    "input normalization used by wh_pos, log_pos, log_pos_mov"
+    if positioners is None:
+        # Grab IPython namespace, recursively find Positioners.
+        positioners = get_all_positioners()
+    elif isinstance(positioners, (Device, Positioner)):
+        # Explore children in case this is a composite Device.
+        positioners = _recursive_positioner_search(positioners)
+    else:
+        # Assume this is a list of Devices.
+        positioners = []
```
Comment (tacaswell) [inline ophyd/commands.py]: I think there is a line missing here?

File: ophyd/commands.py (hunk 2 above)
Comment (danielballan) [inline ophyd/commands.py]: The issue was upstream.

**Pull request merged**